### ISSUE TITLE
[pkg update error: At least one invalid signature was encountered.] coreutils: cp: failed to clone, breaking apt update (was: mv does not work any more?)

### ISSUE THREAD

**Issue by angelog0** [termux/termux-packages#15706]
Description revised by moderator.

### TL;DR

Try https://github.com/termux/termux-packages/issues/15706#issuecomment-1477819107 if your `pkg update` does not work.

### Problem description

* `cp` command fails with message "failed to clone ..."

* `apt-get update` (`pkg update`) fails with message "At least one invalid signature was encountered."

### What steps will reproduce the bug?

```
$ touch $PREFIX/tmp/issue15706-foo
$ cp $PREFIX/tmp/issue15706-foo $PREFIX/tmp/issue15706-bar
cp: failed to clone '/data/data/com.termux/files/usr/tmp/issue15706-bar' from '/data/data/com.termux/files/usr/tmp/issue15706-foo': Permission denied
```

```
$ pkg update
No mirror or mirror group selected. You might want to select one by running 'termux-change-repo'
Checking availability of current mirror:
[*] https://packages-cf.termux.dev/apt/termux-main/: ok
Get:1 https://packages-cf.termux.dev/apt/termux-main stable InRelease [14.0 kB]
Get:2 https://packages-cf.termux.dev/apt/termux-x11 x11 InRelease [14.0 kB]
Err:1 https://packages-cf.termux.dev/apt/termux-main stable InRelease
  At least one invalid signature was encountered.
Err:2 https://packages-cf.termux.dev/apt/termux-x11 x11 InRelease
  At least one invalid signature was encountered.
Fetched 28.0 kB in 3s (8889 B/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
1 package can be upgraded. Run 'apt list --upgradable' to see it.
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages-cf.termux.dev/apt/termux-main stable InRelease: At least one invalid signature was encountered.
W: An error occurred during the signature verification. The repository is not updated and the previous index files will be used. GPG error: https://packages-cf.termux.dev/apt/termux-x11 x11 InRelease: At least one invalid signature was encountered.
W: Failed to fetch https://packages-cf.termux.dev/apt/termux-main/dists/stable/InRelease  At least one invalid signature was encountered.
W: Failed to fetch https://packages-cf.termux.dev/apt/termux-x11/dists/x11/InRelease  At least one invalid signature was encountered.
W: Some index files failed to download. They have been ignored, or old ones used instead.
```

Original description follows.

<hr>

### Problem description

After last update (with coreutils...) it seems that `mv` does not work

### What steps will reproduce the bug?

```
cd
touch foo
mv foo storage/downloads/
mv: failed to clone 'storage/downloads/foo' from 'foo': Permission denied
```

I maintain Termux updated, daily..

The above, yesterday worked..

### System information

termux-info:

```
Termux Variables:
TERMUX_APK_RELEASE=F_DROID
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP_PID=29073
TERMUX_IS_DEBUGGABLE_BUILD=0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://packages.termux.dev/apt/termux-main stable main
# sources.list.d/pointless.list
deb https://its-pointless.github.io/files/21 termux extras
Updatable packages:
All packages up to date
termux-tools version:
1.37.0
Android version:
11
Kernel build information:
Linux localhost 4.9.227-22521289 #2 SMP PREEMPT Wed Jan 11 17:40:49 +07 2023 aarch64 Android
Device manufacturer:
samsung
Device model:
SM-T725
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
```

**Comment by angelog0** [comment]
Also `cp` now has the same defect...

**Comment by alexfarseer** [comment]
Same error for me, I updated `coreutils` and now nothing works. Even `apt update` is not working now. But `busybox` utils seems to be working.

**Comment by TomJo2000** [comment]
It might be best to revert the last commit(s) to the coreutils package until the issue can be root caused.

Probably best to also go ahead and tag @landfillbaby.

**Comment by xtkoba** [comment]
Not reproducible for me.

Can anyone share the `strace -f` output of the failing command? Not sure but it is possible that the command in the new `coreutils` uses some syscalls that are not supported (or not allowed) in some environment.

**Comment by xtkoba** [comment]
Excerpt from `strace` output:

```
openat(AT_FDCWD, "foo", O_WRONLY|O_TRUNC) = 4
ioctl(4, BTRFS_IOC_CLONE or FICLONE, 3) = -1 EACCES (Permission denied)
```

**Comment by xtkoba** [comment]
The error shown in https://github.com/termux/termux-packages/issues/15706#issuecomment-1477266934 itself seems to occur also with coreutils 9.1. Seems like error handling has been changed.

**Comment by xtkoba** [comment]
Reproducible for me with `cp`, but not with `mv`.

A workaround (until fix or revert) seems to be to replace `$PREFIX/bin/cp` with the following shell script:

```sh
#!/bin/sh
exec coreutils --coreutils-prog=cp --reflink=never "$@"
```

Alternatively `cp` from busybox can be used instead, as mentioned above.

**Comment by TomJo2000** [comment]
This should preferably be addressed quickly before most people update to the faulty version of `coreutils`, since if I'm interpreting your previous comment correctly, manual intervention may be needed to resolve the issue.

**Comment by xtkoba** [comment]
Yes, because this breaks `apt` functionality, as you mentioned in another issue.

**Comment by TomJo2000** [comment]
At the risk of causing myself a headache, I'll go do some testing on this issue in order to produce to detailed remediation procedure.

**Comment by xtkoba** [comment]
Fixed for me with `coreutils` version `9.2-1`.

If you still see this type of failure, then please share the `strace -f` output of the failing command.

Keeping this open for now, as the impact of this issue is unmeasurable.

**Comment by truboxl** [comment]
```
curl -OLs https://packages-cf.termux.dev/apt/termux-main/pool/main/c/coreutils/coreutils_9.2-1_aarch64.deb
dpkg -i coreutils_9.2-1_aarch64.deb
```
To restore from a broken state without nuking the entire $PREFIX

**Comment by yirod3721** [comment]
@xtkoba  that is not what happened to me and it is not even remotely related also cp and mv are still working

**Comment by yirod3721** [comment]
also the solution is for aarch64 devices while my device is also arm

**Comment by xtkoba** [comment]
The same goes for arm arch. Just get a working DEB from somewhere (preferably from our primary server) and install it with `dpkg -i`.

If you are a skilled user then you can alternatively do what I wrote in https://github.com/termux/termux-packages/issues/15706#issuecomment-1477277128, or

```bash
test -x /system/bin/cp && ln -sfT /system/bin/cp $PREFIX/bin/cp
```

**Comment by yirod3721** [comment]
@sylirre @xtkoba  i download the new coreutils but i keep getting the same error
and i know that they cant be all be down at the same time but they still wont work

**Comment by yirod3721** [comment]
and what does it mean that they are not signed?

**Comment by yirod3721** [comment]
and i still believe this is a separate issue because the listed answers didnt help me
nor was the same problem present

**Comment by xtkoba** [comment]
@yirod3721 Possibly due to clock skew? It is true that `pkg up` can be broken for various reasons. Please open a NEW issue, with your `termux-info` output.

**Comment by TomJo2000** [comment]
> [and what does it mean that they are not signed?](https://github.com/termux/termux-packages/issues/15706#issuecomment-1477585349)

ordinarily packages have their cryptographic signature verified when being installed from a repository,
this verification may fail if the keys don't match, or in this case, if `apt` isn't behaving as expected due to issues with `coreutils`

You can skip the verification step by installing a package from a local `.deb` files using `dpkg -i /path/to/file.deb`

You can download an updated version of the `coreutils` package from https://packages.termux.dev/apt/termux-main/pool/main/c/coreutils/
As turboxl said earlier, you can use:
```bash
curl -OLs https://packages-cf.termux.dev/apt/termux-main/pool/main/c/coreutils/coreutils_9.2-1_$(uname -m).deb
dpkg -i coreutils_9.2-1_$(uname -m).deb
```
To do this.

**Comment by TomJo2000** [comment]
shoot did not consider that

**Comment by landfillbaby** [comment]
sorry

**Comment by xtkoba** [comment]
Replace `$(uname -m)` with `$(dpkg --print-architecture)`.

**Comment by yirod3721** [comment]
`what do i do now

**Comment by landfillbaby** [comment]
> what do i do now

https://github.com/termux/termux-packages/issues/15706#issuecomment-1477665618

**Comment by xtkoba** [comment]
Summary of #15716: it turns out that the fix in `9.2-1` is insufficient. `9.2-2` has been committed to fix it.

**Comment by landfillbaby** [comment]
again, sorry to everyone that i pushed to master by accident and instead of reverting it i tried to fix it and failed. next time i'll be 100% sure to make a pull request for testing

**Comment by yirod3721** [comment]
it worked thanks a lot i hope i wasnt a nuisance but i dont own a computer or pc so i use it instead

**Comment by Anonymous2716** [comment]
So this v9.2-2 is the fixed ( patched) version? Can now execute `pkg upgrade`? 

I'm still using v9.1 and 'apt list --upgradable' how's v9.2-2 available.

**Comment by pixelb** [comment]
@xtkoba I was wondering if you knew what file system types this issue happened on, or perhaps it's an issue with the kernel logic irrespective of file system. Also I see EACCESS was originally the issue, but there must have been others as your reversed the sense of the check and assume unsupported unless a common error. Do you know what other errnos were set as well as EACCESS though? This is problematic enough that we should handle this in upstream coreutils, so I'll apply some mitigation along this lines.

For my reference perhaps we should avoid the auto reflinking on older kernels.
Perhaps along the lines of https://github.com/coreutils/coreutils/commit/ba5e6885d
I see the kernel in use here is 4.9.227-22521289

**Comment by xtkoba** [comment]
@pixelb `ENOTTY` is implied from #15716, though not enough evidence.

Thanks for the suggestion.

**Comment by pixelb** [comment]
@xtkoba ENOTTY does seem quite random, so your reversed check is probably the best solution.
For reference the first proposed upstream patch for this (equivalent to ours) is https://bugs.gnu.org/bug=62404#8. and any subsequent patches will appear on that bug (until closed).

**Comment by pixelb** [comment]
For my reference:

Actually ENOTTY would be appropriate as it means "inappropriate ioctl" in this context.
Also EACCESS would be appropriate depending on the restricted execution environment that's probably in place for android.

**Comment by xtkoba** [comment]
Note that the patch as of version `9.2-3` is susceptible to `EPERM`:

https://github.com/termux/termux-packages/blob/c71402577e3d30b802cfe80c6894ff5a6f371c78/packages/coreutils/src-copy.c.patch#L12-L14

(which is expected to return `true`.)

I won't *fix* this further on my own for now, as I do not receive any report that this still does not work. Seems like WIP upstream and I would like to wait for a fix there.

As always, if you still see this type of failure, then please share the `strace -f` output of the failing command.

**Comment by hivefi** [comment]
Now here is the issue for my problem. Meanwhile I solved it be re-installing.

**Comment by angelog0** [comment]
Does anyone know why our icons disappeared on GitHub?

**Comment by angelog0** [comment]
Instead of declaring arguments the topic off-topic, answer if you have an answer!

**Comment by 2096779623** [comment]
## cp
```
~ $ touch $PREFIX/tmp/a
~ $ cp $PREFIX/tmp/a $PREFIX/tmp/b
~ $ ls -la $PREFIX/tmp/
total 8
drwx------  2 u0_a185 u0_a185 4096 Mar 25 08:45 .
drwx------ 11 u0_a185 u0_a185 4096 Feb 25 12:33 ..
-rw-------  1 u0_a185 u0_a185    0 Mar 25 08:45 a
-rw-------  1 u0_a185 u0_a185    0 Mar 25 08:45 b
~ $ pkg inf coreutils|grep Version

WARNING: apt does not have a stable CLI interface. Use with caution in scripts.

Version: 9.2-4
~ $
```

## cksum
```
$ echo a > a
~ $ echo b > b
~ $ echo c > c
~ $ sha256sum a b c > a.sha256
~ $ sha1sum a b c > a.sha1
~ $ echo "123" > b
~ $ sha256sum -c a.sha256
a: OK
b: FAILED
c: OK
sha256sum: WARNING: 1 computed checksum did NOT match
~ $ sha1sum -c a.sha1
a: OK
b: FAILED
c: OK
sha1sum: WARNING: 1 computed checksum did NOT match
~ $
```
Work fine.

**Comment by xtkoba** [comment]
Three months have passed since the opening. I would like to close and unpin this issue to make room for another.

Theoretically this can affect some users indefinitely. But hopefully they search the web and find the solution provided here.
